proposals/Layer-1.md

```diff
@@ -0,0 +1,390 @@
+   body. The caught exception is pushed onto the stack.
+
+1. Throws not caught within a function body continue up the call stack, popping
+   call frames, until an enclosing try block is found.
+   
+1. If the call stack is exhausted without any enclosing try blocks, the host VM
+   defines how to handle the uncaught exception.
+
+### Exceptions
+
+An `exception` is an internal construct in WebAssembly that is maintained by the
+host. WebAssembly exceptions (as opposed to host exceptions) are defined by a
+new `exception section` of a WebAssembly module. The exception section is a list
+of exception types, from which exceptions can be created.
+
+Each exception type has a `type signature`. The type signature defines the list
+of values associated with the exception.
+
+Within the module, exception types are identified by an index into the
+[exception index space](#exception-index-space). This index is referred to as
+the `exception tag`. The `tagged exception type` is the corresponding
+exception type refered to by the exception tag.
+
+Exception types can be imported and exported by adding the appropriate entries
+to the import and export sections of the module. All imported/exported exception
+types must be named to reconcile exception tags between modules.
+
+Exception tags are used by:
+
+1. The `except` instruction which creates a WebAssembly instance of the
+   corresponding tagged exception type, and pushes a reference to it onto the
+   stack.
+
+2. The `if_except` instruction that queries an exception to see if it is an
+   instance of the corresponding tagged exception class, and if true it pushes
+   the corresponding values of the exception onto the stack.
+   
+### The exception refernce data type
+
+Data types are extended to have a new `except_ref` type. The representation of
+an exception type is left to the host VM, but its size must be fixed. The actual
+number of bytes it takes to represent any `except_ref` is left to the host VM.
+
+### Try and catch blocks
+
+A _try block_ defines a list of instructions that may need to process exceptions
+and/or clean up state when an exception is thrown. Like other higher-level
+constructs, a try block begins with a `try` instruction, and ends with an `end`
+instruction. That is, a try block is sequence of instructions having the
+following form:
+
+```
+try block_type
+  instruction*
+catch
+  instruction*
+end
+```
+
+A try block ends with a `catch block` that is defined by the list of
+instructions after the `catch` instruction.
+
+Try blocks, like control-flow blocks, have a _block type_. The block type of a
+try block defines the values yielded by the evaluation the try block when either
+no exception is thrown, or the exception is successfully caught by the catch
+block.
+
+In the initial implementation, try blocks may only yield 0 or 1 values.
+
+### Exception creation
+
+A `except` instruction has a single immediate argument, an exception tag.  The
```

Comment:
Separating out exception value creation from throwing is likely to make throw less efficient, because the compiler does not statically know what it is throwing. In the previous design, the compiler statically knew where to read off the handler to jump to (each exn could have a thread-local handler stack attached to it); with the separation it probably requires a hash table lookup.

The separation will also cause problems with future extensions. In particular, when you consider resumption, the type of `throw` will depend on the return type of the exception. That would only work if the `exn_ref` type you propose tracked that result type.

Reply:
I assumed (like return-call) that a throw that also creates could be added as new operator after MVP.

Reply:
Since the combined operation is the simpler and more extensible choice (e.g. avoids the need for an exn_ref type) it seems more natural to go the other way round. Is there an immediate use case for the separated instructions?